### PR TITLE
Allow user to choose sample rate

### PR DESCRIPTION
**Pull request by rtbo**
Following [comment](https://github.com/musescore/MuseScore/issues/18701#issuecomment-2243402697) of @MarcSabatella. It makes really simple for users to fix by themselves the issue #18701.

Only implemented on Linux Alsa.
I can write the code for other platforms, but will be unable to test.

image

- [X] I signed the [CLA](https://musescore.org/en/cla)
- [X] The title of the PR describes the problem it addresses
- [X] Each commit's message describes its purpose and effects, and references the issue it resolves
- [X] If changes are extensive, there is a sequence of easily reviewable commits
- [X] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [X] There are no unnecessary changes
- [X] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Comment by RomanPudashkin** [comment]
>I can write the code for other platforms, but will be unable to test.

Yes, please! You can copy-paste your code for WasapiAudioDriver (although it won't do anything since WASAPI always runs on the system sample rate) and OSXAudioDriver

File: src/appshell/qml/Preferences/internal/CommonAudioApiConfiguration.qml [hunk 1]
```diff
@@ -64,7 +64,25 @@ Item {
                 apiModel.deviceSelected(newValue)
             }
         }
+        
+        ComboBoxWithTitle {
+            id: sampleRate
+
+            title: qsTrc("appshell", "Sample rate:")
```
Comment (RomanPudashkin) [inline src/appshell/qml/Preferences/internal/CommonAudioApiConfiguration.qml:71]: "appshell/preferences"

File: src/appshell/qml/Preferences/internal/CommonAudioApiConfiguration.qml [hunk 2]
```diff
@@ -64,7 +64,25 @@ Item {
...
+            columnWidth: root.columnWidth
+
+            currentIndex: indexOfValue(apiModel.sampleRate)
+            model: apiModel.sampleRateList
+
+            navigation.name: "SampleRateBox"
+            navigation.panel: root.navigation
+            navigation.row: root.navigationOrderStart + 1
```
Comment (RomanPudashkin) [inline src/appshell/qml/Preferences/internal/CommonAudioApiConfiguration.qml:79]: This combobox should probably be placed after "Buffer size:". And after that, the navigation row should be corrected:
navigation.row: root.navigationOrderStart **+ 2**

File: src/appshell/view/preferences/commonaudioapiconfigurationmodel.cpp [hunk 3]
```diff
@@ -77,6 +82,30 @@ void CommonAudioApiConfigurationModel::deviceSelected(const QString& deviceId)
...
+QList<unsigned int> CommonAudioApiConfigurationModel::sampleRateList() const
+{
+    QList<unsigned int> result;
+    std::vector<unsigned int> sampleRates = audioDriver()->availableOutputDeviceSampleRates();
+
+    for (unsigned int sampleRate : sampleRates) {
+        result << sampleRate;
+    }
+
+    std::sort(result.begin(), result.end());
```
Comment (RomanPudashkin) [inline src/appshell/view/preferences/commonaudioapiconfigurationmodel.cpp:99]: Looks like we don't need it...

File: src/framework/audio/internal/platform/lin/linuxaudiodriver.cpp [hunk 4]
```diff
@@ -257,6 +257,50 @@ async::Notification LinuxAudioDriver::availableOutputDevicesChanged() const
...
+}
+
+std::vector<unsigned int> LinuxAudioDriver::availableOutputDeviceSampleRates() const
+{
+    // ALSA API is not of any help to get sample rates supported by the driver.
+    // (snd_pcm_hw_params_get_rate_[min|max] will return 1 to 384000 Hz)
+    // So just returning a sensible hard-coded list.
+    return std::vector<unsigned int> {
+        44100,
+        48000,
```
Comment (RomanPudashkin) [inline src/framework/audio/internal/platform/lin/linuxaudiodriver.cpp:299]: Shouldn't we add 88200 to this list?

File: src/framework/audio/internal/platform/lin/linuxaudiodriver.cpp [hunk 5]
```diff
@@ -257,6 +257,50 @@ async::Notification LinuxAudioDriver::availableOutputDevicesChanged() const
...
 }
 
+unsigned int LinuxAudioDriver::outputDeviceSampleRate() const
+{
+    return s_format.sampleRate;
+}
+
+bool LinuxAudioDriver::setOutputDeviceSampleRate(unsigned int sampleRate)
+{
+    if (s_format.samples == sampleRate) {
```
Comment (RomanPudashkin) [inline src/framework/audio/internal/platform/lin/linuxaudiodriver.cpp:267]: if (s_format.**sampleRate** == sampleRate) {

File: src/framework/audio/internal/platform/lin/linuxaudiodriver.cpp (hunk 4 above)
Comment (cbjeukendrup) [reply src/framework/audio/internal/platform/lin/linuxaudiodriver.cpp:299]: And perhaps 192000 as well? I believe that's not uncommon as well, even my cheap Windows laptop supports it (or at least, gives it as an option...)

**Comment by rtbo** [comment]
There is a bug in MuseSounds. 
Changing the sample rate affects the pitch and speed of the playback.
After switching from 44100Hz to 88200Hz, it plays twice too fast, one octave too high.

Whatever the settings, the app always starts with the correct speed and pitch, however.
So if I restart MS with 88200Hz setting and switch back to 44100Hz, it will be twice too slow, and one octave too low.

With MS Basic synth, there is no issue.

**Comment by DmitryArefiev** [comment]
Tested on LinuxUbuntu22.04, Win10, Mac13.6

On **Linux**, there is already mentioned issue with MuseSounds
On **Windows**,  there is also an issue with MS Basic - when select other sample rates (but it switches back to 44 visually after changing), it affects note pitch and tempo
On **Mac**, when using 48-96kHz (sounds and basic), there are sounds artifacts, especially on 4096 buffer size

**Comment by DmitryArefiev** [comment]
@rtbo Can you hide the sample rate settings on Windows and Mac please? (just leave it for Linux)

**Comment by DmitryArefiev** [comment]
@MarcSabatella Can you try this on your Linux machine please? Does it fix #18701 when switch to 48?

**Review by RomanPudashkin — APPROVED** [review]

**Comment by RomanPudashkin** [comment]
@MarcSabatella this issue is known as https://github.com/musescore/MuseScore/issues/17877. I'm going to fix it separately

**Pull request merged**